xray/handler_test.go

```diff
@@ -27,65 +23,73 @@ func TestNewFixedSegmentName(t *testing.T) {
+// 	os.Setenv("AWS_XRAY_TRACING_NAME", "test_env")
+// 	n := NewFixedSegmentNamer("test")
+// 	assert.Equal(t, "test_env", n.FixedName)
+// 	os.Unsetenv("AWS_XRAY_TRACING_NAME")
+// }
 
 func TestNewDynamicSegmentName(t *testing.T) {
 	n := NewDynamicSegmentNamer("test", "a/b/c")
 	assert.Equal(t, "test", n.FallbackName)
 	assert.Equal(t, "a/b/c", n.RecognizedHosts)
 }
 
-func TestNewDynamicSegmentNameFromEnv(t *testing.T) {
-	os.Setenv("AWS_XRAY_TRACING_NAME", "test_env")
-	n := NewDynamicSegmentNamer("test", "a/b/c")
-	assert.Equal(t, "test_env", n.FallbackName)
-	assert.Equal(t, "a/b/c", n.RecognizedHosts)
-	os.Unsetenv("AWS_XRAY_TRACING_NAME")
-}
+// func TestNewDynamicSegmentNameFromEnv(t *testing.T) {
+// 	os.Setenv("AWS_XRAY_TRACING_NAME", "test_env")
+// 	n := NewDynamicSegmentNamer("test", "a/b/c")
+// 	assert.Equal(t, "test_env", n.FallbackName)
+// 	assert.Equal(t, "a/b/c", n.RecognizedHosts)
+// 	os.Unsetenv("AWS_XRAY_TRACING_NAME")
+// }
 
 func TestHandlerWithContextForRootHandler(t *testing.T) {
-	TestDaemon.Reset()
-	ctx, _ := ContextWithConfig(context.Background(), Config{
-		ServiceVersion: "1.0.0",
-	})
+	ctx, td := NewTestDaemon()
+	defer td.Close()
 
 	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		w.WriteHeader(http.StatusOK)
-		b := []byte(`200 - OK`)
-		w.Write(b)
+		if _, err := w.Write([]byte(`200 - OK`)); err != nil {
+			panic(err)
+		}
 	})
 
 	ts := httptest.NewServer(HandlerWithContext(ctx, NewFixedSegmentNamer("test"), handler))
 	defer ts.Close()
 
-	req := httptest.NewRequest("POST", ts.URL, strings.NewReader(""))
+	req, err := http.NewRequest(http.MethodPost, ts.URL, strings.NewReader(""))
+	if !assert.NoError(t, err) {
+		return
+	}
 	req.Header.Set("User-Agent", "UnitTest")
-	req.Header.Set("X-Forwarded-For", "127.0.0.1")
-
-	_, err := http.DefaultTransport.RoundTrip(req)
-	assert.NoError(t, err)
-
-	s, e := TestDaemon.Recv()
-	assert.NoError(t, e)
 
-	assert.Equal(t, http.StatusOK, s.HTTP.Response.Status)
-	assert.Equal(t, "POST", s.HTTP.Request.Method)
-	assert.Equal(t, ts.URL+"/", s.HTTP.Request.URL)
-	assert.Equal(t, "127.0.0.1", s.HTTP.Request.ClientIP)
-	assert.Equal(t, "UnitTest", s.HTTP.Request.UserAgent)
-	assert.Equal(t, "1.0.0", s.Service.Version)
+	resp, err := http.DefaultClient.Do(req)
+	if !assert.NoError(t, err) {
+		return
+	}
+	resp.Body.Close()
+	assert.Equal(t, http.StatusOK, resp.StatusCode)
+
+	// make sure all connections are closed.
+	ts.Close()
+
+	seg, err := td.Recv()
+	if !assert.NoError(t, err) {
+		return
+	}
+
+	assert.Equal(t, http.StatusOK, seg.HTTP.Response.Status)
+	assert.Equal(t, "POST", seg.HTTP.Request.Method)
+	assert.Equal(t, ts.URL+"/", seg.HTTP.Request.URL)
+	assert.Equal(t, "127.0.0.1", seg.HTTP.Request.ClientIP)
+	assert.Equal(t, "UnitTest", seg.HTTP.Request.UserAgent)
+	assert.Equal(t, "TestVersion", seg.Service.Version)
 }
 
 func TestHandlerWithContextForNonRootHandler(t *testing.T) {
-	TestDaemon.Reset()
```

Comment:
Why did we add 3 functions in 1 function ?

Reply:
I moved TestRootHandler and TestNonRootHandler to `xray/handler_go113_test.go`,
because I have no idea override the configure with `xray.Handler`.

easiest way is using `Configure`, but I can't because of https://github.com/aws/aws-xray-sdk-go/pull/177#discussion_r368186749

another way is using `Server.BaseContext` https://golang.org/pkg/net/http/#Server

```go
	ts.Config.BaseContext = func(_ net.Listener) context.Context {
		return ctx
	}
```

but it is available from Go 1.13, and minimum version of X-Ray SDK is Go 1.9.
So, I moved them to `xray/handler_go113_test.go`, and add build constraint ( `// +build go1.13` )